pkg/clusteragent/autoscaling/workload/patcher.go

```diff
@@ -8,38 +8,116 @@
 package workload
 
 import (
-	"fmt"
-
-	datadoghq "github.com/DataDog/datadog-operator/apis/datadoghq/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/record"
 
 	"github.com/DataDog/datadog-agent/pkg/clusteragent/autoscaling/workload/model"
-
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
-// PatcherAdapter allows a workload patcher to access and modify data available in controller store
-type PatcherAdapter interface {
-	// GetPodAutoscalerFromOwnerRef returns the PodAutoscalerInternal object associated with the given owner reference
-	// NOTE: The returned PodAutoscalerInternal should never be modified
-	GetRecommendations(ns string, ownerRef metav1.OwnerReference) (string, []datadoghq.DatadogPodAutoscalerContainerResources, error)
+// PODPatcher allows a workload patcher to patch a workload with the recommendations from the autoscaler
+type PODPatcher interface {
+	// ApplyRecommendation applies the recommendation to the given pod
+	ApplyRecommendations(pod *corev1.Pod) (bool, error)
 }
 
-type patcherAdapter struct {
-	store *store
+type podPatcher struct {
+	store         *store
+	eventRecorder record.EventRecorder
 }
 
-var _ PatcherAdapter = patcherAdapter{}
+var _ PODPatcher = podPatcher{}
+
+func newPODPatcher(store *store, eventRecorder record.EventRecorder) PODPatcher {
+	return podPatcher{
+		store:         store,
+		eventRecorder: eventRecorder,
+	}
+}
+
+func (pa podPatcher) ApplyRecommendations(pod *corev1.Pod) (bool, error) {
+	autoscaler, err := pa.findAutoscaler(pod)
```

Comment:
Here we're retrieving a reference of the object that is in the store. Is it possible to have a data race if the object is updated by the controller ?

Reply:
No, there's a lock in the store for that, but we may wait on the lock.